### ISSUE TITLE
Proposal: fully explicit, non-exclusive "auto" layouts in WGSL (without createPipelineLayout)

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#4957]
The [`"auto"` layout algorithm](https://gpuweb.github.io/gpuweb/#default-pipeline-layout) currently generates layouts which cannot be shared with other pipelines. The reason for this is that these layouts are derived from static-use analysis of WGSL shaders, which makes the "auto" generated layout sensitive to small changes in the WGSL source (e.g. removing a call to a helper function that uses a binding). This makes the layouts fragile - difficult to share between pipelines, and potentially difficult to debug due to "spooky action at a distance" (making a small code change in one shader makes an entirely different pipeline creation fail).

However, `"auto"` layouts are extremely convenient. It's annoying to have to fall back to explicit layouts as soon as you need to share the layout - which happens quickly in most non-trivial graphics projects - as you suddenly have to duplicate the entire layout on the JS side and store the GPUBindGroupLayout/GPUPipelineLayout objects off somewhere.

To fix this I want to propose two things: #4956, and adding an explicit layout syntax to WGSL. `"auto"` layouts created from this syntax would be very tightly defined in a single `@interface` annotation on the entrypoint(s) of the pipeline (the vertex and fragment layouts would be merged). All bindings referenced from `@interface` would also require explicit filtering/non-filtering and float/unfilterable-float/depth information from #4956.

Rough mock-up with compute:

```rust
@group(0) @binding(0) var<uniform> params: SimParams;
@group(1) @binding(1) var<storage, read> particlesA: Particles;
@group(1) @binding(2) var<storage, read_write> particlesB: Particles;

@interface(params, particlesA, particlesB)
@compute @workgroup_size(64) fn main() { /* ... params ... particlesA ... particlesB ... */ }
```

Bind points could optionally be deferrable until `@interface` time:

```rust
@group() @binding() var<uniform> params: SimParams;
@group() @binding(1) var<storage, read> particlesA: Particles;
@group() @binding(2) var<storage, read_write> particlesB: Particles;

@interface(@group(0) @binding(0) params, @group(1) particlesA, @group(1) particlesB)
@compute @workgroup_size(64) fn main() { /* ... params ... particlesA ... particlesB ... */ }
```

Or better, even batched into entire bind groups with a new type of "struct":

```rust
@group() @binding() var<uniform> params: SimParams;

@group()
bind_group particles {
  @binding(1) var<storage, read> a: Particles;
  @binding(2) var<storage, read_write> b: Particles;
}

@interface(@group(0) @binding(0) params, @group(1) ParticleData)
@compute @workgroup_size(64) fn main() { /* ... params ... particles.a ... particles.b ... */ }
```

Thanks to informal design discussion with teammates after the F2F, especially @dj2 :)

**Comment by kainino0x** [comment]
The `@interface` block defines the layout exactly:
Bindings not in the `@interface` block would not be allowed to be used anywhere in the call tree of the entry point.
Bindings in the `@interface` block are of course always considered statically used.
Binding subtype are not inferred at all, they're required to be explicit.

BTW, immediates (push constants) would also need to be in this interface.

**Comment by alan-baker** [comment]
This feels like a somewhat kludgy workaround for defining the interface as part of the parameter list of the entry point. That would lead to a different coding style, but with `unrestricted_pointer_parameters` it seems possible to today. So your first example would be something like:
```
@compute @workgroup_size(64)
fn main(
  @group(0) @binding(0) params : ptr<uniform, SimParams>,
  @group(0) @binding(1) particlesA : ptr<storage, Particles>,
  @group(0) @binding(2) particlesB : ptr<storage, Particles, read_write>,
  /* any built-in values that are needed */
) {
  someFunction(particlesA, particlesB, params);
}
```
Any helpers would have to be passed the resource by pointer or value (in the case of samplers and textures). It is very Metal-esque.

The `bind_group` is an interesting idea, but I dislike the var of vars aspect as it is proposed. In this parameter style I could see something like:
```
bind_group MyBindGroup {
  @binding(1) particlesA : ptr<storage, Particles>,
  @binding(2) particlesB : ptr<storage, Particles, read_write>,
}

@compute @workgroup_size(64)
fn main(
  @group(0) @binding(0) params : ptr<uniform, SimParams>,
  @group(1) bg1 : MyBindGroup,
  /* built-in values */
) {
  someFunction(bg1.particleA, bg1.particlesB, params);
} 
```

Admittedly this would require some changes to the alias analysis in the spec, but I think we have to re-examine that anyways for binding arrays.

**Comment by magcius** [comment]
Agreed that I would like to see function parameters used here, and I prefer it to `@interface` syntax. I was thinking of proposingt before, but my concern was that you couldn't easily pass it as a pointer to a function, but it sounds like that's taken care of with `unrestricted_pointer_parameters`, so I'm all for it now.

I think I would prefer having a struct be used for grouping together sets of bindings, rather than introducing a new `bind_group` syntax. We already use structs for function inputs and outputs between stages, so it seems natural to support them as taking in bindings too.

My one request is that if something is explicitly part of a function parameter, that it not be subject to the static-use rules -- it's just considered part of the function's "ABI" in that case.

**Comment by kainino0x** [comment]
A struct of pointers sounds good to me. I just didn't know how to write out a proposal with pointers so I did it in the simplest way possible.

I thought it would be annoying to pass bindings down into every function call, but if we allow pointers to bindings in structs then that's not so bad as they can be grouped. Spitballing some things that could potentially help with this:
- non-flat structs would be useful so you can pass substructs into functions
- laying out the bindings arbitrarily in the structs (not necessarily grouped by `group`) might be useful
- "aliasing" bindings might be useful, seems possible if they are all pointers

```rust
struct SamplerTexture0 {
  @group(0) @binding(0) sampler: ...;
  @group(1) @binding(0) texture0: ...;
}
struct SamplerTexture1 {
  @group(0) @binding(0) sampler: ...; // sampler is from the same bind point!
  @group(1) @binding(1) texture1: ...; // but use a different texture
}
struct ParticlesBindGroup {
  @binding(1) particlesA : ptr<storage, Particles>,
  @binding(2) particlesB : ptr<storage, Particles, read_write>,
}
struct Bindings {
  st0: SamplerTexture0, // groups 0 and 1 set inside
  st1: SamplerTexture1, // groups 0 and 1 set inside
  @group(2) particles: ParticlesBindGroup, // group 2 set outside
}

@compute @workgroup_size(64)
fn main(bindings: Bindings) {
  someFunction(bindings.st0);
  otherFunction(bindings.particles);
}
```

**Comment by jimblandy** [comment]
CC @mighdoll

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐11‐19-Minutes#proposal-fully-explicit-non-exclusive-auto-layouts-in-wgsl-without-createpipelinelayout--issue-4957">WGSL 2024-11-19 Minutes</a></summary>

* KG: We didn’t have much time to go over this, so we need more time to give it more thought.
* JB: I do like the direction though, especially for explaining to new people.
* JB: It’s this thing that you’re always imagining in your head, but now you would get tooling support.
* AB: It seems like JSP is saying they don’t want to make this about static-use, like the proposal proposes.
* JB: I wonder if we could encourage the WESL people to experiment with this?
* KG: We can probably just @them on the issue? (or if you’re reading this: hi!)
* DN: So what’s the impetus to this?
* KG: It’s nice when you e.g. have the same vertex binding layouts between two shaders.
* ds: It sounds like JSP is asking for declaration in the @interface to count as static-use for the purposes of auto-layout. 
* **(no resolution today)**

</details>

**Comment by magcius** [comment]
> (or if you’re reading this: hi!)

Hi! Yes, my intent is that if you declare something in a function argument, it's always seen as statically used.

My argument is that static use seems like it's mostly valuable if you have a large shader module with a lot of entry points; having all items be in the auto layout is too wide a scope; you could pick up resources you don't want in your shader. So static use is a nice way to "pare down" the ABI to what you're using.

However, that comes with a bit of a footgun, where the resulting binding group is based on the compiler's usage, so you have to "think like a compiler" in order to use it most effectively, and this was one of the more visible issues I've helped new users with. It was common back in the day to see people comment out a `textureSample` line and get errors from the automatic binding layouts.

With this new syntax, you have a way of explicitly saying which resources you're using, so it makes sense that if it's part of a function argument, it should always be part of the binding layout. That logic feels fairly easy to understand without having to think like a compiler.

**Comment by mighdoll** [comment]
Hi! 

Awesomeness. Yep, we're listening now. Thanks for tagging us. 

I've brought it to the WESL discord to make folks aware and discuss. (We're enthusiastic about helping on things like this that can be prototyped in user space.)

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐12‐03-Minutes#proposal-fully-explicit-non-exclusive-auto-layouts-in-wgsl-without-createpipelinelayout--issue-4957">WGSL 2024-12-03 Minutes</a></summary>

* JB: Seems like folks want a single source of expressing the binding layout group. 
* KG: There’s still a use case to specify it in the API side.
* KG: so far this is just a colloquial proposal?
* ds: Yeah we would need a syntax proposal here. I think we were hoping to get feedback from the WESL folks, but we could also try to come up with something first.
* JB: I suspect WESL would appreciate guidance e.g. “Try AB’s proposal”.
* ds: I like @interface but that’s just what we came up with initially.
* AB: I think there’s a lot of options. E.g. is it good to pass things around as a struct. 
* MW: Have we had concrete requests from devs here? Like, are devs often using explicit layouts?
* ds: Well, part of the problem is that auto-layouts aren’t re-useable, so that might push people away “artificially”. 
* KG: So is this two issues: 1) fully-explicit and 2) non-exclusive
* ds: You kinda need one for the other, because making changes in one shader would cause it to stop matching with others. 
* DN: So if you want shareable layouts, will you want the same blob in both shader sources?
* ds: Maybe? Or define multiple	shaders in the same wgsl file.
* JP: Is it true that the explicit layout thing is just syntactic sugar vs adding dummy vars?
* ds: Yes, but if you change one pipeline layout, it might change another one “spookily”. 
* JP: But authors could create compatible layouts to make them continue to match?
* JB: [It’s kind of more than strictly syntactic sugar]
* KG: So we want explicit here to “fix” things in place to prevent “spooky at a distance” behavior, but we think this is a blocker for non-excl auto layouts.
* ds: That’s my understanding from talking with KN
* AB: Would be interesting to take to WESL: About composability of layouts, having to rebind things. At least you could say “here’s a layout” to pass round. 
* KG: do we have an @explicit attrib for this kind of thing yet? Or that’s what we’re asking for? Because if we did, we could e.g. Warn if you ask for an `auto` layout from something where not everything’s marked.
* DN: Could have a dev feature to lint for whether there’s something in a helper/util func that accesses something not in some allowlist of entrypoint “expected bindings”.
* **Resolved: Sounds cool, but please make a spec proposal. FYI: We kind of like the struct approach.**
* JB: I would really like to have people try things before speccing things. I think it would be great to have real devs give feedback on real use of this. I don’t know how fast the WESL folk would be able to prototype this, but it would be great to get some kind of user feedback here. I don’t want to Require us to get this feedback before proceeding, but I think we should try to ask for this. But e.g. Origin Trials feels too high-overhead for this. 
* KG: I think this would need to be prototyped in an implementation because this is lifting a restriction we have, not polyfillable. But an impl can just add a behind-a-flag prototype for this.
* AB: Not to put MW on the spot, but this is kinda like the transition from ~GLSL to ~C++-via-MTL. Did people like that?
* MW: I agree with the overall direction, just not sure how long it will take, and not sure how.
* DN: I want to go look at the Slang research again, because they address interface reuse, so that comparison would be nice.
* 

</details>

**Comment by mighdoll** [comment]
Sorry, we've been talking about this but on different channels (mostly on discord). Some bits now, more to follow:

We've been talking about several scenarios on how to try out 4957 style binding structs: 
- Typed reflection. There the wgsl/wesl shader layouts are reflected into host code (e.g. TypeScript) as typed objects by a continuously code generating build tool like vite. The programmer uses a TypeScript `import` statement to load the generated layout.
- Untyped reflection. The wgsl/wesl shader layouts are reflected into host code live at runtime, and without type information. The programmer makes calls to `wesl_link()` with their source, and is returned as not just a wgsl string, but also untyped reflection objects.
- Polyfill. Untyped reflection, but via monkey patching the WebGPU API. 
- Typed injection. Shader layouts are authored in host code e.g. TypeScript and injected into .wesl. The build tool continuously generates the WESL code. The programmer uses a WESL `import` statement to load the generated bindings. 
- (I imagine we'll defer polyfill for a bit - the most important thing is to try the layout ideas. Probably start with one of the reflection ideas, and then also demo injection..)

The typed approaches allow the host language server (e.g. TypeScript) to statically validate if bind groups match their layout on the type-level. 

With injection, the host code is the source of truth, which gives maximum control to the application to handle dynamic scenarios. [TypeGPU](https://typegpu.com) and [Use.GPU](https://usegpu.live/) have found this to be valuable. We're imagining both approaches should co-exist. Some projects will prefer to define layouts in the shaders for simplicity and some in host code for flexibility.

As far as the WGSL syntax for describing shareable binding group layouts:
- What's supported by WGSL with `unrestricted_pointer_parameters`? We have the technical ability to 'lower' WESL to WGSL of course, but are uncertain what's in or coming in wrt to the structures and pointers described in 4957.
- We've been workshopping an idea with the (placeholder) name 'shareable-auto'. The rough idea is to give developers an alternate auto layout algorithm that bases the layout on all binding definitions included in the shader module passed into `create*Pipeline()`, rather than on bindings used by the entry functions. Stay tuned for more details, we'd like to discuss the design idea.
- We like the idea of sharing group/binding indices between host and shader code by name, so that programmers don't have to manually keep indices in sync. Most of the reflection/injection approaches achieve that along with nice typing. But it'd be nice to cook up a way to share indices by name that can work w/o tooling too.
- Of course, some of these constructs with group/binding indices don't work as well for shared libraries, but that's a problem for another day, and it's fine for apps regardless. Hand waving that things are probably fine for libraries... Probably we want apps to control the indices and inject them into libraries, so we'd probably want a mechanism like this in apps plus an injection mechanism for libraries (e.g. `@extern`/`@link`).

Next steps
- @iwoplaza and friends at [Software Mansion](https://swmansion.com/) have been leading the explorations here, and are planning a prototype for the TypeScript WESL tool chain.. (@iwoplaza is one of the deepest thinkers in the community on shader/TypeScript integration.)

**Comment by Kangz** [comment]
`shareable-auto` is interesting but it seems to go lead people to have one shader module per entry-point when we have been trying to favor the opposite (so that code parsing and validation is shared). It does however have the advantage to be an extremely small change to the API surface, and somewhat simple code change in the WebGPU implementations.

The bindings structs that @kainino0x suggested above would make it easy for shader authors to group things they know will be used together and have explicit control of what get included in the entrypoint. This way there is less "action at a distance" possible, which has been another important design constraint so far (and the reason why auto BGLs cannot be shared between pipelines).

**Comment by jimblandy** [comment]
Moz has warm fuzzy feelings about: https://github.com/gpuweb/gpuweb/issues/4957#issuecomment-2471702174

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2025‐01‐28-Minutes#proposal-fully-explicit-non-exclusive-auto-layouts-in-wgsl-without-createpipelinelayout--issue-4957">WGSL 2025-01-28 Minutes</a></summary>

* KG: Still in design phase of this one. WESL folks asking for feedback on a particular idea that they called sharable auto. Mozilla quite liked KNs comment about structs with pointers. Big concern from CW about shareable auto is that it will push folks towards separating modules into different files instead of having in a single file because you cant' have colliding bindings like having something different bound into two different entry points wouldn't work if it had a static use.
* DS: The suggestion from WESL seemed a step backward from what Kai suggested.  The need to split into separate files is less nice. Building it into  struct seems nicer.
* KG: Interesting to get the feel in the room for what folks like before a proposal is writing. Not sure how high priority this is for folks. Suspect we're in implementation and implementation issue phase and think thats the right thing to do. This is a nice to have that we'll timeslice in when we have time.
* IP: I'm helping WESL with the ts integration and have been discussing internally. Shareable auto was an alternate proposal and not something that we prefer over bindings in structs. Just an alternate proposal that would decrease the amount of surface and supports the bindings statements. Feedback was valid and understand that splitting into multiple files is more of a step backwards then what was initially proposed. Working to figure out how we can do this in userland and what approaches we can do like polyfill. How we can create a better user experience for creating bindings.
* JB: The kind of experimentation that WESL does is extremely valuable. Would be wonderful to have experiments to say we tried and didn't work out as we expected. That stuff is gold.

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2025‐02‐18-Minutes#proposal-fully-explicit-non-exclusive-auto-layouts-in-wgsl-without-createpipelinelayout--issue-4957">WGSL 2025-02-18 Minutes</a></summary>

* **Still waiting on feedback.**

</details>

**Comment by jimblandy** [comment]
<details>
<summary>minutes from WGSL committee meeting 2025-10-28</summary>
<p>JimB: How does this proposal related to #5353? Should we put off discussing it until #5353 is resolved?
ds: 5353 just provides more information to auto layouts, so it's related but doesn't solve the issue completely.</p>
</details>

The committee also [discussed this issue in more detail](https://github.com/gpuweb/gpuweb/issues/5353#issuecomment-3458480195) in the context of #5353.
